ds0_updater.py

```diff
@@ -5,13 +5,36 @@
 import logger
 import conf
 
+# keep it super simple for now because we don't really implement --allowed-keys yet.
+def verify_latest_commit():
+    allowed_keys = ds0_utils.allowed_keys()
+
+    # ensure the latest commit is signed and verified
+    while True:
+        for key in allowed_keys:
+            logger.exe('gpg --import ' + key['file'], expectError=True)
+        output = logger.exe('git log --pretty="format:%G?" --show-signature HEAD^..HEAD')
+
+        if "Can't check signature" in output[0]:
+            logger.info('gpg keys cleared on startup. Trying again...')
+            continue
+
+        verified = False
+        for key in allowed_keys:
+            if key['rsa_check'] in output[0] and key['signature_check'] in output[0]:
+                verified = True
+        if not verified:
+            logger.error('Scripts using a non-signed commit. Please ensure commit is valid.')
+            logger.error('    If it was a missed signature, feel free to open a ticket at https://github.com/riptano/ComboAMI.')
+        break
+
 # Update the AMI codebase if it's its first boot
 if not conf.get_config("AMI", "CompletedFirstBoot"):
     # check if a specific commit was requested
     force_commit = ds0_utils.required_commit()
 
     # update the repo
-    logger.exe('git pull')
+    logger.exe('git pull ' + ds0_utils.repository())
```

Comment:
Using the hash as a repo/refspec separator looks fine to me. If npm is doing it, it's probably sensible. I also agree it's a much more intuitive interface than the --forcecommit flag.

It would be very nice if --repository could accept any refspec. Right now, it can only accept a branch name because git doesn't know the commit-hashes for the custom-repo. Rather than git pulling (which cannot accept a commit-hash from the remote) Can we git-remote-add (which will grab the commit-history with hashes), followed by a git reset --hard <refspec> to modify the workspace which can accept a branch, tag, or commit-hash from either repo. I'm fine hard-coding the name of the remote since we can only support a single one anyway.

It would also be good for --forcecommit and the refspec piece of the --repository option to be synonyms. They do the same thing, which is reset the workingspace to a particular refspec. I think we don't want to drop --forcecommit for backwards compatibility reasons, but could the repository function:

 1. Return the offical repo by default
 2. Check both --forcecommit and the refspec piece of --repository, return whichever is set, and if both are set return the refspec set in --repository and log a warning.

With these changes, I'm +1 on this PR and will mege/test it. Thanks a ton for doing this and for being so responsive to feedback.